flaky test: TestAgent_ReloadConfigAndKeepChecksStatus

https://circleci.com/gh/hashicorp/consul/139535

```
--- FAIL: TestAgent_ReloadConfigAndKeepChecksStatus (7.23s)
    writer.go:29: 2020-03-10T15:42:21.307Z [WARN]  TestAgent_ReloadConfigAndKeepChecksStatus: bootstrap = true: do not enable unless necessary
    writer.go:29: 2020-03-10T15:42:21.308Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server.raft: initial configuration: index=1 servers="[{Suffrage:Voter ID:8ad7be10-d3a2-41df-f7f1-1343955446ad Address:127.0.0.1:10113}]"
    writer.go:29: 2020-03-10T15:42:21.308Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server.raft: entering follower state: follower="Node at 127.0.0.1:10113 [Follower]" leader=
    writer.go:29: 2020-03-10T15:42:21.309Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server.serf.wan: serf: EventMemberJoin: Node-8ad7be10-d3a2-41df-f7f1-1343955446ad.dc1 127.0.0.1
    writer.go:29: 2020-03-10T15:42:21.309Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server.serf.lan: serf: EventMemberJoin: Node-8ad7be10-d3a2-41df-f7f1-1343955446ad 127.0.0.1
    writer.go:29: 2020-03-10T15:42:21.309Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server: Handled event for server in area: event=member-join server=Node-8ad7be10-d3a2-41df-f7f1-1343955446ad.dc1 area=wan
    writer.go:29: 2020-03-10T15:42:21.309Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server: Adding LAN server: server="Node-8ad7be10-d3a2-41df-f7f1-1343955446ad (Addr: tcp/127.0.0.1:10113) (DC: dc1)"
    writer.go:29: 2020-03-10T15:42:21.310Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: Started DNS server: address=127.0.0.1:10122 network=udp
    writer.go:29: 2020-03-10T15:42:21.310Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: Started DNS server: address=127.0.0.1:10122 network=tcp
    writer.go:29: 2020-03-10T15:42:21.310Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: Started HTTP server: address=127.0.0.1:10123 network=tcp
    writer.go:29: 2020-03-10T15:42:21.310Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: started state syncer
    writer.go:29: 2020-03-10T15:42:21.310Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: Started gRPC server: address=127.0.0.1:10114 network=tcp
    writer.go:29: 2020-03-10T15:42:21.360Z [WARN]  TestAgent_ReloadConfigAndKeepChecksStatus.server.raft: heartbeat timeout reached, starting election: last-leader=
    writer.go:29: 2020-03-10T15:42:21.360Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server.raft: entering candidate state: node="Node at 127.0.0.1:10113 [Candidate]" term=2
    writer.go:29: 2020-03-10T15:42:21.360Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus.server.raft: votes: needed=1
    writer.go:29: 2020-03-10T15:42:21.360Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus.server.raft: vote granted: from=8ad7be10-d3a2-41df-f7f1-1343955446ad term=2 tally=1
    writer.go:29: 2020-03-10T15:42:21.360Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server.raft: election won: tally=1
    writer.go:29: 2020-03-10T15:42:21.360Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server.raft: entering leader state: leader="Node at 127.0.0.1:10113 [Leader]"
    writer.go:29: 2020-03-10T15:42:21.360Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server: cluster leadership acquired
    writer.go:29: 2020-03-10T15:42:21.360Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server: New leader elected: payload=Node-8ad7be10-d3a2-41df-f7f1-1343955446ad
    writer.go:29: 2020-03-10T15:42:21.361Z [DEBUG] connect.ca.consul: consul CA provider configured: id=07:80:c8:de:f6:41:86:29:8f:9c:b8:17:d6:48:c2:d5:c5:5c:7f:0c:03:f7:cf:97:5a:a7:c1:68:aa:23:ae:81 is_primary=true
    writer.go:29: 2020-03-10T15:42:21.362Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server.connect: initialized primary datacenter CA with provider: provider=consul
    writer.go:29: 2020-03-10T15:42:21.362Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.leader: started routine: routine="federation state anti-entropy"
    writer.go:29: 2020-03-10T15:42:21.362Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.leader: started routine: routine="federation state pruning"
    writer.go:29: 2020-03-10T15:42:21.362Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.leader: started routine: routine="CA root pruning"
    writer.go:29: 2020-03-10T15:42:21.362Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus.server: Skipping self join check for node since the cluster is too small: node=Node-8ad7be10-d3a2-41df-f7f1-1343955446ad
    writer.go:29: 2020-03-10T15:42:21.362Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server: member joined, marking health alive: member=Node-8ad7be10-d3a2-41df-f7f1-1343955446ad
    writer.go:29: 2020-03-10T15:42:21.362Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server: federation state anti-entropy synced
    writer.go:29: 2020-03-10T15:42:21.363Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Skipping remote check since it is managed automatically: check=serfHealth
    writer.go:29: 2020-03-10T15:42:21.364Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: Synced node info
    writer.go:29: 2020-03-10T15:42:21.364Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: Synced service: service=webserver1
    writer.go:29: 2020-03-10T15:42:21.364Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Check in sync: check=service:webserver1
    writer.go:29: 2020-03-10T15:42:21.364Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Node info in sync
    writer.go:29: 2020-03-10T15:42:21.364Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Service in sync: service=webserver1
    writer.go:29: 2020-03-10T15:42:21.364Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Check in sync: check=service:webserver1
    writer.go:29: 2020-03-10T15:42:21.414Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Check status updated: check=service:webserver1 status=passing
    writer.go:29: 2020-03-10T15:42:22.414Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Check status updated: check=service:webserver1 status=passing
    writer.go:29: 2020-03-10T15:42:22.708Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Skipping remote check since it is managed automatically: check=serfHealth
    writer.go:29: 2020-03-10T15:42:22.708Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Node info in sync
    writer.go:29: 2020-03-10T15:42:22.708Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Service in sync: service=webserver1
    writer.go:29: 2020-03-10T15:42:22.708Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: Synced check: check=service:webserver1
    writer.go:29: 2020-03-10T15:42:22.708Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Node info in sync
    writer.go:29: 2020-03-10T15:42:22.708Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Service in sync: service=webserver1
    writer.go:29: 2020-03-10T15:42:22.708Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Check in sync: check=service:webserver1
    writer.go:29: 2020-03-10T15:42:23.415Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Check status updated: check=service:webserver1 status=passing
    writer.go:29: 2020-03-10T15:42:24.416Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Check status updated: check=service:webserver1 status=passing
    writer.go:29: 2020-03-10T15:42:25.417Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Check status updated: check=service:webserver1 status=passing
    writer.go:29: 2020-03-10T15:42:26.418Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Check status updated: check=service:webserver1 status=passing
    writer.go:29: 2020-03-10T15:42:27.419Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Check status updated: check=service:webserver1 status=passing
    writer.go:29: 2020-03-10T15:42:28.420Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus: Check status updated: check=service:webserver1 status=passing
    retry.go:121: assertions.go:261: 
        	Error Trace:	agent_test.go:3463
        	            				retry.go:130
        	            				asm_amd64.s:1357
        	Error:      	Not equal: 
        	            	expected: "critical"
        	            	actual  : "passing"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-critical
        	            	+passing
        	Messages:   	check {"service:webserver1" {}} is wrong
        
        
    writer.go:29: 2020-03-10T15:42:28.532Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: Requesting shutdown
    writer.go:29: 2020-03-10T15:42:28.532Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server: shutting down server
    writer.go:29: 2020-03-10T15:42:28.532Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus.leader: stopping routine: routine="federation state anti-entropy"
    writer.go:29: 2020-03-10T15:42:28.532Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus.leader: stopping routine: routine="federation state pruning"
    writer.go:29: 2020-03-10T15:42:28.532Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus.leader: stopping routine: routine="CA root pruning"
    writer.go:29: 2020-03-10T15:42:28.532Z [WARN]  TestAgent_ReloadConfigAndKeepChecksStatus.server.serf.lan: serf: Shutdown without a Leave
    writer.go:29: 2020-03-10T15:42:28.532Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus.leader: stopped routine: routine="federation state pruning"
    writer.go:29: 2020-03-10T15:42:28.532Z [DEBUG] TestAgent_ReloadConfigAndKeepChecksStatus.leader: stopped routine: routine="CA root pruning"
    writer.go:29: 2020-03-10T15:42:28.532Z [WARN]  TestAgent_ReloadConfigAndKeepChecksStatus.server.serf.wan: serf: Shutdown without a Leave
    writer.go:29: 2020-03-10T15:42:28.532Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus.server.router.manager: shutting down
    writer.go:29: 2020-03-10T15:42:28.532Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: consul server down
    writer.go:29: 2020-03-10T15:42:28.532Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: shutdown complete
    writer.go:29: 2020-03-10T15:42:28.532Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: Stopping server: protocol=DNS address=127.0.0.1:10122 network=tcp
    writer.go:29: 2020-03-10T15:42:28.532Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: Stopping server: protocol=DNS address=127.0.0.1:10122 network=udp
    writer.go:29: 2020-03-10T15:42:28.532Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: Stopping server: protocol=HTTP address=127.0.0.1:10123 network=tcp
    writer.go:29: 2020-03-10T15:42:28.532Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: Waiting for endpoints to shut down
    writer.go:29: 2020-03-10T15:42:28.532Z [INFO]  TestAgent_ReloadConfigAndKeepChecksStatus: Endpoints down
```

I have a suspicion that #7345 restored a racy behavior that had previously been fixed, while it was fixing a different thing that regressed.

--- a/agent/agent_test.go
+++ b/agent/agent_test.go
@@ -3443,35 +3443,27 @@ func TestAgent_ReloadConfigAndKeepChecksStatus(t *testing.T) {
 	t.Parallel()
 	dataDir := testutil.TempDir(t, "agent") // we manage the data dir
 	defer os.RemoveAll(dataDir)
-	waitDurationSeconds := 1
 	hcl := `data_dir = "` + dataDir + `"
 		enable_local_script_checks=true
 		services=[{
 		  name="webserver1",
-		  check{name="check1",
-		  args=["true"],
-		  interval="` + strconv.Itoa(waitDurationSeconds) + `s"}}
-		]`
+		  check{id="check1", ttl="30s"}
+		}]`
 	a := NewTestAgent(t, t.Name(), hcl)
 	defer a.Shutdown()
 
-	// We wait until the checks are passing to do the reload.
-	retry.Run(t, func(r *retry.R) {
-		gotChecks := a.State.Checks(nil)
-		for id, check := range gotChecks {
-			require.Equal(r, "passing", check.Status, "check %q is wrong", id)
-		}
-	})
+	require.NoError(t, a.updateTTLCheck(structs.NewCheckID("check1", nil), api.HealthPassing, "testing agent reload"))
+
+	// Make sure check is passing before we reload.
+	gotChecks := a.State.Checks(nil)
+	require.Equal(t, 1, len(gotChecks), "Should have a check registered, but had %#v", gotChecks)
+	for id, check := range gotChecks {
+		require.Equal(t, "passing", check.Status, "check %q is wrong", id)
+	}
 
 	c := TestConfig(testutil.Logger(t), config.Source{Name: t.Name(), Format: "hcl", Data: hcl})
 	require.NoError(t, a.ReloadConfig(c))
 	// After reload, should be passing directly (no critical state)
-	for id, check := range a.State.Checks(nil) {
-		require.Equal(t, "passing", check.Status, "check %q is wrong", id)
-	}
-	// Ensure to take reload into account event with async stuff
-	time.Sleep(time.Duration(100) * time.Millisecond)
-	// Of course, after a sleep, should be Ok too
 	for id, check := range a.State.Checks(nil) {
 		require.Equal(t, "passing", check.Status, "check %q is wrong", id)
 	}

--- a/agent/agent_test.go
+++ b/agent/agent_test.go
@@ -3455,18 +3455,7 @@ func TestAgent_ReloadConfigAndKeepChecksStatus(t *testing.T) {
 	a := NewTestAgent(t, t.Name(), hcl)
 	defer a.Shutdown()
 
-	// Initially, state is critical during waitDurationSeconds seconds
-	retry.Run(t, func(r *retry.R) {
-		gotChecks := a.State.Checks(nil)
-		require.Equal(r, 1, len(gotChecks), "Should have a check registered, but had %#v", gotChecks)
-		for id, check := range gotChecks {
-			require.Equal(r, "critical", check.Status, "check %q is wrong", id)
-		}
-	})
-	c := TestConfig(testutil.Logger(t), config.Source{Name: t.Name(), Format: "hcl", Data: hcl})
-	a.ReloadConfig(c)
-	time.Sleep(time.Duration(waitDurationSeconds) * time.Second)
-	// It should now be passing
+	// We wait until the checks are passing to do the reload.
 	retry.Run(t, func(r *retry.R) {
 		gotChecks := a.State.Checks(nil)
 		for id, check := range gotChecks {
@@ -3474,6 +3463,7 @@ func TestAgent_ReloadConfigAndKeepChecksStatus(t *testing.T) {
 		}
 	})
 
+	c := TestConfig(testutil.Logger(t), config.Source{Name: t.Name(), Format: "hcl", Data: hcl})
 	require.NoError(t, a.ReloadConfig(c))
 	// After reload, should be passing directly (no critical state)
 	for id, check := range a.State.Checks(nil) {

